plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/LazyConnectionFactory.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class LazyConnectionFactory
+        implements ConnectionFactory
+{
+    private final ConnectionFactory delegate;
+
+    public LazyConnectionFactory(ConnectionFactory delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        return new LazyConnection(() -> delegate.openConnection(session));
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        delegate.close();
+    }
+
+    private static final class LazyConnection
+            extends ForwardingConnection
+    {
+        private final SqlSupplier<Connection> connectionSupplier;
+        private volatile Connection connection;
+
+        public LazyConnection(SqlSupplier<Connection> connectionSupplier)
+        {
+            this.connectionSupplier = requireNonNull(connectionSupplier, "connectionSupplier is null");
+        }
+
+        @Override
+        protected Connection getDelegate()
+                throws SQLException
+        {
+            if (connection == null) {
+                synchronized (this) {
+                    if (connection == null) {
+                        connection = connectionSupplier.get();
+                    }
+                }
+            }
+            return connection;
+        }
+
+        @Override
+        public void close()
+                throws SQLException
+        {
+            if (connection != null) {
+                connection.close();
+            }
+        }
+    }
```

Comment:
Add `unwrap` and `isWrapperFor` so that `lazyConn.unwrap(Connection.class)` returns `lazyConn`